policy/approval/approve.go

```diff
@@ -251,29 +281,40 @@ func (r *Rule) IsApproved(ctx context.Context, prctx pull.Context) (bool, string
 	return false, msg, nil
 }
 
-func (r *Rule) filteredCandidates(ctx context.Context, prctx pull.Context) ([]*common.Candidate, error) {
+func (r *Rule) filteredCandidates(ctx context.Context, prctx pull.Context) ([]*common.Candidate, []*common.Candidate, error) {
 	candidates, err := r.Options.GetMethods().Candidates(ctx, prctx)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get approval candidates")
+		return nil, nil, errors.Wrap(err, "failed to get approval candidates")
 	}
 
 	sort.Stable(common.CandidatesByCreationTime(candidates))
 
 	if r.Options.IgnoreEditedComments {
 		candidates, err = r.filterEditedCandidates(ctx, prctx, candidates)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
 	if r.Options.InvalidateOnPush {
 		candidates, err = r.filterInvalidCandidates(ctx, prctx, candidates)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
-	return candidates, nil
+	var allowedCandidates []*common.Candidate
+	var discardedCandidates []*common.Candidate
+
+	for _, c := range candidates {
+		if len(c.DiscardedBecause) == 0 {
```

Comment:
While this works, I'm wondering if it would be cleaner to pass this distinction down to `filterInvalidCandidates` and `filterEditedCandidates`, having them return two lists (and not setting a reason). This function could then aggregate the results, attach the reason based on which function discarded the candidate, possibly returning a `[]*common.DismissedReview` instead of a []*common.Candidate`.

Ideally, I'd like to keep valid and discarded candidates separate, rather than having some functions that return a mixed list (where you have to check the reason) and some that return separate lists (where it's expected that one list sets the reason and the other doesn't.) This could also allow removing the `DiscardedBecause` field from `common.Candidate`.

Reply:
this is what i had at first, and changed it, will change it back.